akka-stream/src/main/scala/akka/stream/impl/FanIn.scala

```diff
@@ -241,7 +241,7 @@ private[akka] abstract class FanIn(val settings: ActorFlowMaterializerSettings,
   protected def fail(e: Throwable): Unit = {
     if (settings.debugLogging)
       log.debug("fail due to: {}", e.getMessage)
-    inputBunch.cancel()
+    nextPhase(completedPhase)
```

Comment:
this ensures that if we are in the WaitingForUpstreamSubscription state then we will wait for all subscriptions before we finish. (nextPhase does not immediately take effect if we are in WaitingForUpstreamSubscription)